src/truss.cpp

```diff
@@ -0,0 +1,242 @@
+ *
+ * \param graph The input graph.
+ * \param support The support of each edge, specified as a vector where the
+ * indices stand for the edges, and support is (initially!) the number of
+ * triangles the edge occurs in. This is used as a workspace and is changed,
+ * then destroyed.
+ * \param truss This will indicate the highest k-truss each edge occurs in.
+ * \return Error code.
+ *
+ *
+ * Time complexity: It should be O(|E|^1.5). See Algorithm 2 in:
+ * Wang, Jia, and James Cheng. "Truss decomposition in massive networks."
+ * Proceedings of the VLDB Endowment 5.9 (2012): 812-823.
+ *
+ *
+ * \example examples/simple/igraph_truss.c
+ */
+void trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss){
+
+  // Get max possible value = max entry in support.
+  int max = 0;
+  for (long int i = 0; i < igraph_vector_int_size(support); ++i) {
+    int value = VECTOR(*support)[i];
+    if (max < value) { max = value; }
+  }
+
+  // The vector of levels. Each level of the vector is a set of edges initially
+  // at that level of support, where support is # of triangles the edge is in.
+  vector< unordered_set<long int> > vec(max + 1);
+
+  // Initialize completed edges.
+  igraph_vector_bool_t completed;
+  igraph_vector_bool_init(&completed, igraph_ecount(graph));
+
+  // Add each edge to its appropriate level of support.
+  for (long int i = 0; i < igraph_vector_int_size(support); ++i) {
+    vec[VECTOR(*support)[i]].insert(i);  // insert edge i into its support level
+  }
+
+  // Record the trussness of edges at level 0.
+  unordered_set<long int>::iterator it;
+  for (it = vec[0].begin(); it != vec[0].end(); ++it){
+    VECTOR(*truss)[*it] = 0;
+    VECTOR(completed)[*it] = 1;  // this edge is already completed
+  }
+
+  // Initialize variables needed below.
+  igraph_integer_t fromVertex, toVertex, e1, e2;
+  igraph_vector_t fromNeighbors, toNeighbors, q1, q2, commonNeighbors;
+  igraph_vector_init(&fromNeighbors, 0);
+  igraph_vector_init(&toNeighbors, 0);
+  igraph_vector_init(&commonNeighbors, 0);
+
+  // Move through the levels, one level at a time, starting at first level.
+  for (int level = 1; level <= max; ++level){
+    while (!vec[level].empty()){
+      long int seed = *vec[level].begin();  // pull out the first edge
+      vec[level].erase(seed);  // remove the first element
+
+      // Find neighbors.
+      igraph_edge(graph, seed, &fromVertex, &toVertex);
+      igraph_neighbors(graph, &fromNeighbors, fromVertex, IGRAPH_ALL);
+      igraph_neighbors(graph, &toNeighbors, toVertex, IGRAPH_ALL);
+      q1 = fromNeighbors;
+      q2 = toNeighbors;
+
+      if (igraph_vector_size(&q1) > igraph_vector_size(&q2)){
+        // case: #fromNeighbors > #toNeigbors, so make q1 the smaller set.
+        q1 = toNeighbors;
+        q2 = fromNeighbors;
+      }
+
+      // Sort the neighbors.
+      igraph_vector_sort(&q1);
+      igraph_vector_sort(&q2);
```

Comment:
Lots of extra legwork is being done here because you are querying and sorting the neighbors of a single vertex many times during the iterations. Consider using an `igraph_adjlist_t` data structure in this function, followed by `igraph_adjlist_sort()` to speed things up - basically the adjacency list will contain the pre-sorted neighbor vectors so you can simply query them later.